File: midonet-cluster/src/main/java/org/midonet/cluster/rest_api/models/IPSecPolicy.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.rest_api.models;

import java.net.URI;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.google.common.base.MoreObjects;

import org.midonet.cluster.data.ZoomClass;
import org.midonet.cluster.data.ZoomEnum;
import org.midonet.cluster.data.ZoomEnumValue;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Neutron;
import org.midonet.cluster.rest_api.ResourceUris;
import org.midonet.cluster.rest_api.neutron.models.IPSecAuthAlgorithm;
import org.midonet.cluster.rest_api.neutron.models.IPSecEncryptionAlgorithm;
import org.midonet.cluster.rest_api.neutron.models.IPSecPfs;

@ZoomClass(clazz = Neutron.IPSecPolicy.class)
public class IPSecPolicy extends UriResource {

    @ZoomField(name = "id")
    public UUID id;

    @ZoomField(name = "tenant_id")
    public String tenantId;

    @ZoomField(name = "name")
    public String name;

    @ZoomField(name = "description")
    public String description;

    @ZoomField(name = "transform_protocol")
    public TransformProtocol transformProtocol;

    @ZoomField(name = "auth_algorithm")
    public IPSecAuthAlgorithm authAlgorithm = IPSecAuthAlgorithm.SHA1;

    @ZoomField(name = "encryption_algorithm")
    public IPSecEncryptionAlgorithm encryptionAlgorithm;

    @ZoomField(name = "encapsulation_mode")
    public EncapsulationMode encapsulationMode;

    @ZoomField(name = "lifetime_value")
    public Integer lifetimeValue;

    @ZoomField(name = "lifetime_units")
    public String lifetimeUnits;

    @ZoomField(name = "pfs")
    public IPSecPfs pfs;

    @ZoomEnum(clazz = Neutron.IPSecPolicy.TransformProtocol.class)
    public enum TransformProtocol {
        @ZoomEnumValue("ESP") ESP,
        @ZoomEnumValue("AH") AH,
        @ZoomEnumValue("AH_ESP") AH_ESP
    }

    @ZoomEnum(clazz = Neutron.IPSecPolicy.EncapsulationMode.class)
    public enum EncapsulationMode {
        @ZoomEnumValue("TUNNEL") TUNNEL,
        @ZoomEnumValue("TRANSPORT") TRANSPORT
    }

    public URI getUri() {
        return absoluteUri(ResourceUris.IPSEC_POLICIES, id);
    }

    @Override
    @JsonIgnore
    public void create() {
        if (null == id) {
            id = UUID.randomUUID();
        }
    }

    @JsonIgnore
    public void update(IPSecPolicy from) {
        id = from.id;
    }

    public String toString() {
        return MoreObjects.toStringHelper(this)
            .omitNullValues()
            .add("id", id)
            .add("tenantId", tenantId)
            .add("name", name)
            .add("description", description)
            .add("transformProtocol", transformProtocol)
            .add("encryptionAlgorithm", encryptionAlgorithm)
            .add("encapsulationMode", encapsulationMode)
            .add("lifetimeValue", lifetimeValue)
            .add("lifetimeUnits", lifetimeUnits)
            .add("pfs", pfs)
            .toString();
    }
}

File: python-midonetclient/src/midonetclient/ipsec_policy.py
# Copyright (c) 2015 Midokura Europe SARL, All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from midonetclient import resource_base
from midonetclient import vendor_media_type

class IPSecPolicy(resource_base.ResourceBase):

    media_type = vendor_media_type.APPLICATION_IPSEC_POLICY_JSON

    def __init__(self, uri, dto, auth):
        super(IPSecPolicy, self).__init__(uri, dto, auth)

    def get_id(self):
        return self.dto['id']

    def get_tenant_id(self):
        return self.dto['tenantId']

    def get_name(self):
        return self.dto['name']

    def get_description(self):
        return self.dto['description']

    def get_transform_protocol(self):
        return self.dto['transformProtocol']

    def get_auth_algorithm(self):
        return self.dto['authAlgorithm']

    def get_encryption_algorithm(self):
        return self.dto['encryptionAlgorithm']

    def get_encapsulation_mode(self):
        return self.dto['encapsulationMode']

    def get_lifetime_value(self):
        return self.dto['lifetimeValue']

    def get_lifetime_units(self):
        return self.dto['lifetimeUnits']

    def get_pfs(self):
        return self.dto['pfs']

    def name(self, name):
        self.dto['name'] = name
        return self

    def description(self, description):
        self.dto['description'] = description
        return self

    def transform_protocol(self, transform_protocol):
        self.dto['transformProtocol'] = transform_protocol
        return self

    def auth_algorithm(self, auth_algorithm):
        self.dto['authAlgorithm'] = auth_algorithm
        return self

    def encryption_algorithm(self, encryption_algorithm):
        self.dto['encryptionAlgorithm'] = encryption_algorithm
        return self

    def encapsulation_mode(self, encapsulation_mode):
        self.dto['encapsulationMode'] = encapsulation_mode
        return self

    def lifetime_value(self, lifetime_value):
        self.dto['lifetimeValue'] = lifetime_value
        return self

    def lifetime_units(self, lifetime_units):
        self.dto['lifetimeUnits'] = lifetime_units
        return self

    def pfs(self, pfs):
        self.dto['pfs'] = pfs
        return self